### DISCUSSION TITLE
What do you think about my Sound extension?

### DISCUSSION THREAD

**Discussion by GiovanniSalmeri** [datenstrom/community#866] [Ask a question] [answered]
Hello everybody! I have published a new [Sound](https://github.com/GiovanniSalmeri/yellow-sound) extension. It is analogous to the [Video](https://github.com/GiovanniSalmeri/yellow-video) extension, but while the latter leverages the standard HTML5 video widget for embedding media files, Sound uses its own widget (written from scratch), which allows customisation, responsiveness etc. I have made my best in order to follow the guidelines for accessibility, but I haven't done a real test from this point of view. Besides audio files, Sound can also embed playlists and several third-party services. More sound widgets can be used in a page and all behaves as expected.

In order to get the informations of the audio file (author, title ecc.), I have written and included a small library which reads audio tags. Several similar libraries do exist, but none seemed to me really fit for this purpose (either too limited, or too general, or client-side ecc.), so I have written all from scratch, and so discovered that audio tags are really a nightmare 😱. I hope that the final result works well enough (audio tags are read from mp3, flac, opus, ogg/vorbis, m4a).

A final word about similar exensions already available:

+ [Audio](https://github.com/schulle4u/yellow-extensions-schulle4u/tree/main/audio) by @schulle4u uses the standard HTML5 widget and is perfectly adequate and in my opinion preferable when other features are not needed
+ [Soundcloud](https://github.com/schulle4u/yellow-extensions-schulle4u/tree/main/soundcloud) by @schulle4u embeds tracks from the service Soundcloud
+ [Jouele](https://github.com/kodersha/yellow-jouele) by @kodersha is similar in concept to Sound (custom widget + audio tags), with a minimalist look

In PHP code  there are still a lot of TODOs (which perhaps will become DONTFIX's 😊), but anyway I have preferred to publish the extension as it is. Any comments are welcome. Thank you!

Edit: the extension behaves a little erratically with local audio files and the embedded server of Yellow (`php yellow.php serve`). I suspect that this is the consequence of security measures in browsers that prevent access to local files. Anyway, these problems disappear in a proper installation, so I did not investigate further.

**Comment by schulle4u** [comment, marked as answer]
Thanks Giovanni, this looks amazing! 👍

However supporting so many patterns at once might have some drawbacks, e.g. when the URL differs from the expected format. I've tested this with a mp3 radio stream that doesn't have an `.mp3` extension at the end of the URL, and it broke the shortcut. You can append a fake extension to avoid this, e.g. `[sound http://example.com/8000/stream?.mp3]`

**Reply by GiovanniSalmeri** [reply]
Thank you very much for your feedback!

Regarding the many patterns supported: yes, I too wonder whether I have put too much meat on the fire (as we say in Italian). I like a lot the idea that each program, or each extension, must do "one thing well", but it isn't always easy to understand what "one" thing is from the viewpoint of a user or webmaster. In this case I have supposed that embedding audio tracks, from whatever source, is "one" thing, but I do not know if the result is functional and clear enough. Anyway, this is the first attempt.

Regarding in particular URLs without extension: you have spotted the first TODO of this extension! 😊 I had thought about a simple mechanism that reads the headers when no extension is present, I will try to code it as soon as possible. For the time being, thank you for the suggestion about a fake extension, I wouldn't ever thought of it!

**Reply by GiovanniSalmeri** [reply]
Now it should work with URLs without an extension at the end. I wonder whether the code could be simpler (for example allowing remote URL only for Internet radios, which should be the most common case, or splitting the extension in two, one for audio tracks and one for radios). Simplicity is difficult to achieve! 😅

**Reply by schulle4u** [reply]
Thanks, I'll give it a try. As for splitting extensions I would indeed vote for a separate radio extension. This not only allows correct handling of the common stream types (Icecast/Shoutcast/HLS etc), you could also integrate some specific functions offered by streaming servers, e.g listener stats, song history and so on. Especially Icecast has a nice JSON API to access such informations. I once did some basic testing with a pure php icecast extension, but wasn't satisfied with the results and my knowledge for writing a JS version is still very limited. 🙂

**Reply by GiovanniSalmeri** [reply]
> This not only allows correct handling of the common stream types (Icecast/Shoutcast/HLS etc), you could also integrate some specific functions offered by streaming servers, e.g listener stats, song history and so on.

This sounds interesting! Actually I know very little about Internet radios, and the only thing that Sound currently does when a radio stream is detected is to omit timeline and total time and to take the name of the radio from the header `icy-name` (which unfortunately, as I have learnt with some tests, is often left blank or badly set). So something better would be very useful, I will explore as soon as possible :smiley:

**Reply by PetersOtto** [reply]
I have a Raspberry Pi Zero with HiFiBerry mini Amp running as an internet radio in my office and in our kitchen. The Pi runs headless and can be controlled via internet (mobilephone, tablet, PC ..).

The Pi runs LAMP and the mpc/mpd client (https://www.musicpd.org/clients/mpc/). Starting point for my developments was this tutorial: https://fabianflu.ch/raspberry-pi-webradio-extended-edition/ It is written in german ...

Beside the radio streams it is possible to play Spotify. For this I use raspotify: https://github.com/dtcooper/raspotify

I'm not sure if this fits to your project but maybe yes :)

**Reply by GiovanniSalmeri** [reply]
@PetersOtto Thank you very much for the references! Soon or later I will study them... :-)
